extensions/ql-vscode/src/local-queries.ts

```diff
@@ -7,384 +7,575 @@ import {
-    databaseItem: dbItem,
-    label: dbItem.name,
-    description: dbItem.language,
-  }));
-  /**
-   * Databases that were selected in the quick pick menu.
-   */
-  const quickpick = await window.showQuickPick<DatabaseQuickPickItem>(
-    quickPickItems,
-    { canPickMany: true, ignoreFocusOut: true },
-  );
-  if (quickpick !== undefined) {
-    // Collect all skipped databases and display them at the end (instead of popping up individual errors)
-    const skippedDatabases = [];
-    const errors = [];
-    for (const item of quickpick) {
-      try {
-        await compileAndRunQuery(
-          qs,
-          qhm,
-          databaseUI,
-          localQueryResultsView,
-          queryStorageDir,
-          false,
-          uri,
-          progress,
-          token,
-          item.databaseItem,
+    // If possible, only show databases with the right language (otherwise show all databases).
+    const queryLanguage = await findLanguage(this.cliServer, uri);
+    if (queryLanguage) {
+      filteredDBs = this.databaseManager.databaseItems.filter(
+        (db) => db.language === queryLanguage,
+      );
+      if (filteredDBs.length === 0) {
+        void showAndLogErrorMessage(
+          `No databases found for language ${queryLanguage}. Please add a suitable database to your workspace.`,
         );
-      } catch (e) {
-        skippedDatabases.push(item.label);
-        errors.push(getErrorMessage(e));
+        return;
       }
     }
-    if (skippedDatabases.length > 0) {
-      void extLogger.log(`Errors:\n${errors.join("\n")}`);
-      void showAndLogWarningMessage(
-        `The following databases were skipped:\n${skippedDatabases.join(
-          "\n",
-        )}.\nFor details about the errors, see the logs.`,
-      );
+    const quickPickItems = filteredDBs.map<DatabaseQuickPickItem>((dbItem) => ({
+      databaseItem: dbItem,
+      label: dbItem.name,
+      description: dbItem.language,
+    }));
+    /**
+     * Databases that were selected in the quick pick menu.
+     */
+    const quickpick = await window.showQuickPick<DatabaseQuickPickItem>(
+      quickPickItems,
+      { canPickMany: true, ignoreFocusOut: true },
+    );
+    if (quickpick !== undefined) {
+      // Collect all skipped databases and display them at the end (instead of popping up individual errors)
+      const skippedDatabases = [];
+      const errors = [];
+      for (const item of quickpick) {
+        try {
+          await this.compileAndRunQuery(
+            false,
+            uri,
+            progress,
+            token,
+            item.databaseItem,
+          );
+        } catch (e) {
+          skippedDatabases.push(item.label);
+          errors.push(getErrorMessage(e));
+        }
+      }
+      if (skippedDatabases.length > 0) {
+        void extLogger.log(`Errors:\n${errors.join("\n")}`);
+        void showAndLogWarningMessage(
+          `The following databases were skipped:\n${skippedDatabases.join(
+            "\n",
```

Comment:
The popup box is not very large. I don't think it can handle more than 3-4 lines. Have you tried it when there are lots of skipped DBs? Maybe better to concatenate by a `,`.

Reply:
This is not new code, although the diff view might make it look like it is. I only changed `compiledAndRunQuery()` and its callees.